File: ios/Classes/Methods/ConversationsMethods.swift
import Flutter
import TwilioConversationsClient

public class ConversationsMethods {
    public static func createConversation(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let friendlyName = arguments["friendlyName"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing 'friendlyName' parameter", details: nil))
        }

        let conversationOptions: [String: Any] = [
            TCHConversationOptionFriendlyName: friendlyName
        ]

        SwiftTwilioConversationsPlugin.instance?.client?.createConversation(options: conversationOptions, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                let conversationDict = Mapper.conversationToDict(conversation)
                flutterResult(Mapper.encode(conversationDict))
            } else {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error creating conversation with friendlyName '\(friendlyName)': \(String(describing: result.error))", details: nil))
            }
        })
    }
    
    public static func getMyConversations(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        let myConversations =  SwiftTwilioConversationsPlugin.instance?.client?.myConversations()
        let dict = Mapper.conversationsToDict(myConversations)
        flutterResult(Mapper.encode(dict))
    }

    public static func getConversation(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let conversationSidOrUniqueName = arguments["conversationSidOrUniqueName"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing 'conversationSidOrUniqueName' parameter", details: nil))
        }

        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSidOrUniqueName, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                flutterResult(Mapper.encode(Mapper.conversationToDict(conversation)))
            } else {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid or uniqueName '\(conversationSidOrUniqueName)'", details: nil))
            }
        })
    }

}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/ConversationsMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.google.gson.Gson
import com.twilio.conversations.CallbackListener
import com.twilio.conversations.Conversation
import com.twilio.conversations.ErrorInfo
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin

object ConversationsMethods {
    private const val TAG = "ConversationsMethods"

    fun getConversation(call: MethodCall, result: MethodChannel.Result) {
        val conversationSidOrUniqueName = call.argument<String>("conversationSidOrUniqueName")
                ?: return result.error("ERROR", "Missing 'conversationSidOrUniqueName'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSidOrUniqueName, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                result.success(Gson().toJson(Mapper.conversationToMap(conversation)))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun createConversation(call: MethodCall, result: MethodChannel.Result) {
        val friendlyName = call.argument<String>("friendlyName")
                ?: return result.error("ERROR", "Missing 'friendlyName'", null)

        try {
            TwilioConversationsPlugin.client?.createConversation(friendlyName, object : CallbackListener<Conversation?> {
                override fun onSuccess(conversation: Conversation?) {
                    if (conversation == null) {
                        TwilioConversationsPlugin.debug("${call.method} => onError: Conversation null")
                        result.error(call.method, "Error creating conversation: Conversation null", null)
                        return
                    }
                    TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                    val jsonMap = Gson().toJson(Mapper.conversationToMap(conversation))
                    result.success(jsonMap)
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun getMyConversations(call: MethodCall, result: MethodChannel.Result) {
        GlobalScope.launch {
            val myConversations = TwilioConversationsPlugin.client?.myConversations
            var conversationsSynchronized = false

            while (!conversationsSynchronized) {
                conversationsSynchronized = true

                val convoStatuses = myConversations?.map { it.synchronizationStatus }
                convoStatuses?.forEach {
                    conversationsSynchronized = (conversationsSynchronized && (it == Conversation.SynchronizationStatus.ALL))
                }

                delay(100)
            }

            launch(Dispatchers.Main) {
                TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                val jsonMap = Gson().toJson(Mapper.conversationsToMap(myConversations))
                result.success(jsonMap)
            }
        }
    }
}